### PR TITLE
Add GD32L233R-EVAL board support

### PR DESCRIPTION
**Pull request by cameled**
GD32L233R-EVAL board based on Coretex M23 MCU(include System Timer).

### CODE REVIEW

File: west.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ manifest:
       groups:
         - hal
     - name: hal_gigadevice
-      revision: dd0e0322474462b58059e6fedaf1d67d2a0864d0
```
Comment (gmarull) [inline west.yml:71]: `pull/29/head`

File: boards/arm/gd32l233r_eval/gd32l233r_eval_defconfig [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 BrainCo Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SOC_SERIES_GD32L23X=y
+CONFIG_SOC_GD32L233=y
+CONFIG_BOARD_GD32L233R_EVAL=y
+
+
```
Comment (soburi) [inline boards/arm/gd32l233r_eval/gd32l233r_eval_defconfig:8]: Duplicated blank line.
Should be better to remove it.

File: include/zephyr/dt-bindings/clock/gd32l23x-clocks.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#define GD32_CLOCK_SRAM0      GD32_CLOCK_CONFIG(AHB1EN, 2U)
+#define GD32_CLOCK_FMC        GD32_CLOCK_CONFIG(AHB1EN, 4U)
+#define GD32_CLOCK_CRC        GD32_CLOCK_CONFIG(AHB1EN, 6U)
+#define GD32_CLOCK_SRAM1      GD32_CLOCK_CONFIG(AHB1EN, 7U)
+#define GD32_CLOCK_GPIOA      GD32_CLOCK_CONFIG(AHB1EN, 17U)
+#define GD32_CLOCK_GPIOB      GD32_CLOCK_CONFIG(AHB1EN, 18U)
+#define GD32_CLOCK_GPIOC      GD32_CLOCK_CONFIG(AHB1EN, 19U)
+#define GD32_CLOCK_GPIOD      GD32_CLOCK_CONFIG(AHB1EN, 20U)
+#define GD32_CLOCK_GPIOF      GD32_CLOCK_CONFIG(AHB1EN, 22U)
+
```
Comment (soburi) [inline include/zephyr/dt-bindings/clock/gd32l23x-clocks.h:39]: It seems not to define AHB2-related definitions.
Is it your intent?

https://github.com/zephyrproject-rtos/hal_gigadevice/pull/29/files#diff-5ac2452bfb4597667a852e7df7626f3b49421c625cb82b0c7654590018f74622R292-R293

File: include/zephyr/dt-bindings/reset/gd32l23x.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#define GD32_RESET_CTC        GD32_RESET_CONFIG(APB1RST, 30U)
+
+/* APB2 peripherals */
+#define GD32_RESET_SYSCFG     GD32_RESET_CONFIG(APB2RST, 0U)
+#define GD32_RESET_CMP        GD32_RESET_CONFIG(APB2RST, 1U)
+#define GD32_RESET_ADC        GD32_RESET_CONFIG(APB2RST, 9U)
+#define GD32_RESET_TIMER8     GD32_RESET_CONFIG(APB2RST, 11U)
+#define GD32_RESET_SPI0       GD32_RESET_CONFIG(APB2RST, 12U)
+#define GD32_RESET_USART0     GD32_RESET_CONFIG(APB2RST, 14U)
+
```
Comment (soburi) [inline include/zephyr/dt-bindings/reset/gd32l23x.h:65]: It does not define AHB2-related here too.
Add if needed.
https://github.com/zephyrproject-rtos/hal_gigadevice/pull/29/files#diff-5ac2452bfb4597667a852e7df7626f3b49421c625cb82b0c7654590018f74622R348-R349

File: west.yml (hunk 1 above)
Comment (cameled) [reply west.yml:71]: Thanks! I made a screenshot of this, always forget how to do it;-).

File: boards/arm/gd32l233r_eval/gd32l233r_eval_defconfig (hunk 2 above)
Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval_defconfig]: Thanks! remove the duplicated line.

File: include/zephyr/dt-bindings/clock/gd32l23x-clocks.h (hunk 3 above)
Comment (cameled) [reply include/zephyr/dt-bindings/clock/gd32l23x-clocks.h:39]: Added, thanks!

File: include/zephyr/dt-bindings/reset/gd32l23x.h (hunk 4 above)
Comment (cameled) [reply include/zephyr/dt-bindings/reset/gd32l23x.h:69]: Added, thanks!

**Review by nandojve — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
@nandojve @gmarull @soburi please take another look

**Review by soburi — DISMISSED** [review]

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		zephyr,flash-controller = &fmc;
+	};
+
+	soc {
+		/* SRAM0(16K) and SRAM1(16K) address is continuous. */
+		sram: memory@20000000 {
+			compatible = "mmio-sram";
+			reg = <0x20000000 DT_SIZE_K(32)>;
+		};
+	};
```
Comment (gmarull) [inline boards/arm/gd32l233r_eval/gd32l233r_eval.dts:29]: question: why isn't this part of soc files?

Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:29]: There doesn't have a real sram in here. But we introduce a trick usage at gd32l233r_eval board to get a larger sram.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+	leds {
```
[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:31]: I'm not familiar with the key.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts (hunk 5 above)
Comment (gmarull) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:29]: If SRAM0/1 has to be used as a single block, then let's define it this way in soc dts files.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts (hunk 6 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:31]: image

Btw, board image seems to have a watermark, meaning copyright, I guess. This needs to be fixed.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts (hunk 5 above)
Comment (nandojve) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:29]: I`m fine with that if we keep SRAM0, SRAM1 and a combined region SRAM.

Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts]: done, move `sram` definition to `gd32l233rc` dts.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts (hunk 6 above)
[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:23]: done, use a no copyright one instead.

File: boards/arm/gd32l233r_eval/doc/index.rst [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+********
+
+The GD32L233R-EVAL board is a hardware platform that enables design and debug
+of the GigaDevice GD32L233 Cortex-M23 Low Power MCU.
+
+The GD32RCT6 features a single-core ARM Cortex-M4F MCU which can run up
+to 64-MHz with flash accesses zero wait states, 256kB of Flash, 32kB of
+SRAM and 59 GPIOs.
+
+.. image:: img/gd32l233r_eval.jpg
```
[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/doc/index.rst:16]: Sorry, I mixed these.

[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/doc/index.rst:16]: Hope I can get a new one soon. I have adjust my gd32l233r-eval board for custom usage...

[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/doc/index.rst:16]: This will be delay two or three days.

[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/doc/index.rst:16]: @gmarull I have upload the gd32l233r-eval board iamge.

File: boards/arm/gd32l233r_eval/gd32l233r_eval.dts (hunk 6 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:23]: so can wakeup/tamper be added as `gpio-keys`?

[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:23]: OK, I will add it tonight.

[the remark this answers is not in the document]
Comment (cameled) [reply boards/arm/gd32l233r_eval/gd32l233r_eval.dts:23]: done.

**Review by gmarull — DISMISSED** [review]

**Comment by cameled** [comment]
ci failure caused by others.

**Review by nandojve — DISMISSED** [review]

**Comment by carlescufi** [comment]
@cameled please rebase

**Review by nandojve — CHANGES_REQUESTED** [review]
Fix commit message:

```diff
-soc: arm: gigadeivce: add gd32l23x series
+soc: arm: gigadevice: add gd32l23x series

```

**Comment by cameled** [comment]
> 

Thanks! I have add a step to check the commit message spelling.

**Comment by cameled** [comment]
> Fix commit message:
> 
> ```diff
> -soc: arm: gigadeivce: add gd32l23x series
> +soc: arm: gigadevice: add gd32l23x series
> ```

Thanks! I have add a step to check the commit message spelling.

**Review by nandojve — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Comment by cameled** [comment]
A boundary test for SRAM0(16K) and SRAM1(16K). 

If I write `uint16_t` low byte to SRAM0 and high byte to SRAM1, it works too.

```
#include <zephyr/kernel.h>

static K_THREAD_STACK_DEFINE(m_stack, KB(20));
static struct k_thread m_thread_data;

static void thread_handler(void)
{
	uint8_t data[KB(16)];
	uint8_t *addr1 = &data[KB(11) + 0x147];
	uint8_t *addr2 = &data[KB(11) + 0x148];
	uint16_t *addr3 = (uint16_t *)&data[KB(11) + 0x147];

	printk("addr1: %p\n", addr1);
	printk("addr2: %p\n", addr2);
	printk("addr3: %p\n", addr3);

	*addr3 = 0xFF00;
	printk("*addr1: 0x%02x\n", *addr1);
	printk("*addr2: 0x%02x\n", *addr2);
}

void main(void)
{
	k_thread_create(&m_thread_data, m_stack,
			K_THREAD_STACK_SIZEOF(m_stack),
			(k_thread_entry_t)thread_handler,
			NULL, NULL, NULL, K_PRIO_PREEMPT(8), 0, K_NO_WAIT);
}
```
Output:
```
addr1: 0x20003fff
addr2: 0x20004000
addr3: 0x20003fff
*addr1: 0x00
*addr2: 0xff
```

**Pull request merged**